Add initial semantic check, and fix some web tests

dlaliberte opened the pull request:
This PR adds initial support for semantic checks for form fields.  
Currently, only a single field value can be checked.
This is to partially address #3452. 

Here is what it looks like:

![image](https://github.com/GoogleChrome/chromium-dashboard/assets/570125/27321f2f-7410-4c24-9996-6ed38645ae10)


Several miscellaneous errors that are visible after running webtest, but without causing test failures (reported in #2977), are fixed here as well.  Errors about the unknown feature_links urls remain.

Also, in the unit tests, I couldn't figure out how to make a change to the value and await the change.  Seems that `await component.updateComplete` should do it, but it didn't work.  I'll work on this in future PRs.

dlaliberte commented:
I've added the fieldValues array to each of the form fields so that each form field can get the values of other fields.  I still need to add a way to notify a field when another field  that it depends on is changed to rerun the checks.

jrobbins commented on client-src/elements/chromedash-form-field.js:
```diff
@@ -2,52 +2,66 @@ import {LitElement, html, nothing} from 'lit';
+
+
+  getValue() {
+    return !this.value && this.fieldProps.initial ?
```
Could this be `this.value || this.fieldProps.iniitial`?
Is there a logical difference?

dlaliberte replied:
I copied that code that used to be in renderWidget, to be on the safe side.  I'm not sure what it should be, but what you suggest seems reasonable.  The logic might involve whether a value is 0 or null.  We should allow 0 values, so perhaps it should be:

  this.value != null ? this.value : this.fieldProps.initial

jrobbins replied:
OK, that looks like it would work.

jrobbins commented on client-src/elements/chromedash-form-field.js:
```diff
@@ -81,6 +95,7 @@ export class ChromedashFormField extends LitElement {
     } else {
       fieldValue = e.target.value;
     }
+    this.value = fieldValue; // Should we do this?
```
Is this comment resolved now?  It seems like you are doing it.

dlaliberte replied:
The comment was a reminder to myself to be careful about doing this.  We currently don't change the value property, but is there a reason why we should not do so?  It is a dynamic property, so there could potentially be an update loop, for example.  But since this is happening in the event handler for the last update, perhaps the loop is avoided?  I'm not sure.  Seems to work fine with the text fields I am checking now, but maybe other types of fields will behave differently.

jrobbins replied:
I suppose the only way is to test each field type.

dlaliberte replied:
Added TODO to the comment.  I'll keep an eye on this.

jrobbins commented on client-src/elements/chromedash-form-field.js:
```diff
@@ -91,15 +106,49 @@ export class ChromedashFormField extends LitElement {
+          </span>`;
+
+        const formFieldElement = this.renderRoot.querySelector(`#id_${this.name}`);
+        if (formFieldElement && formFieldElement.setCustomValidity &&
```
Th first two parts of this condition could be combined to a little shorter as `formFieldElement?.setCustomValidity`

dlaliberte replied:
Right.  Good point.

jrobbins reviewed (commented):
Some initial comments.

dlaliberte commented:
> I've added the fieldValues array to each of the form fields so that each form field can get the values of other fields. I still need to add a way to notify a field when another field that it depends on is changed to rerun the checks.

Doing checks with multiple fields is working now.  The origin trials desktop start and end milestones must be in order:

![image](https://github.com/GoogleChrome/chromium-dashboard/assets/570125/83cb959b-ecd1-4f76-8c09-39102b4230e3)

I will add a visual test of this, since it will be cumbersome to test in unit tests because, currently, the code maintains a registry of all form field components at the app level.

jrobbins commented on client-src/js-src/cs-client.js:
```diff
@@ -103,24 +103,24 @@ class ChromeStatusClient {
   }
 
-  doGet(resource, body) {
+  async doGet(resource, body) {
```
Is there a reason why `async` is needed or is this just to clarify that each of these functions returns a promise?

dlaliberte replied:
This is needed to avoid *some* of the errors after webtest runs.  More remain, but I am not sure how to fix them.

jrobbins reviewed (commented):
Why are there so many new and modified screenshots?  There seem to be more than the ones created by your new tests.

dlaliberte commented:
I can't seem to reliably delete a feature, let alone take a screenshot of the result.  So I'll avoid feature deletion for now, but resetting the database is useful for avoiding the accumulation of test features.

dlaliberte commented:
> Why are there so many new and modified screenshots? There seem to be more than the ones created by your new tests.

I replaced a '-' with '/' in the screenshot paths, so they will be more readable. I could have done this in a separate PR, but oh well.
